src/coreComponents/constitutive/solid/porosity/PorosityBase.cpp

```diff
@@ -71,45 +65,31 @@ void PorosityBase::allocateConstitutiveData( dataRepository::Group & parent,
 
 void PorosityBase::postProcessInput()
 {
-  getWrapper< array1d< real64 > >( viewKeyStruct::referencePorosityString() )
-    .setApplyDefaultValue( m_defaultReferencePorosity );
+  getExtrinsicData< extrinsicMeshData::porosity::referencePorosity >().
+    setApplyDefaultValue( m_defaultReferencePorosity );
 }
 
-
-void PorosityBase::saveConvergedState() const
+void PorosityBase::scaleReferencePorosity( arrayView1d< real64 const > scalingFactors ) const
 {
   localIndex const numE = numElem();
-  localIndex const numQ = numQuad();
 
-  arrayView2d< real64 const > newPorosity = m_newPorosity;
-  arrayView2d< real64 >       porosity_n  = m_porosity_n;
+  arrayView1d< real64 > referencePorosity = m_referencePorosity;
 
   forAll< parallelDevicePolicy<> >( numE, [=] GEOSX_HOST_DEVICE ( localIndex const k )
   {
-    for( localIndex q = 0; q < numQ; ++q )
-    {
-      porosity_n[k][q] = newPorosity[k][q];
-    }
+    referencePorosity[k] *= scalingFactors[k];
   } );
 }
 
-void PorosityBase::initializeState() const
+void PorosityBase::saveConvergedState() const
 {
-  localIndex const numE = numElem();
-  localIndex const numQ = numQuad();
-
-  arrayView2d< real64 const > newPorosity     = m_newPorosity;
-  arrayView2d< real64 >       porosity_n      = m_porosity_n;
-  arrayView2d< real64 >       initialPorosity = m_initialPorosity;
+  m_porosity_n.setValues< parallelDevicePolicy<> >( m_newPorosity.toViewConst() );
+}
 
-  forAll< parallelDevicePolicy<> >( numE, [=] GEOSX_HOST_DEVICE ( localIndex const k )
-  {
-    for( localIndex q = 0; q < numQ; ++q )
-    {
-      porosity_n[k][q]      = newPorosity[k][q];
-      initialPorosity[k][q] = newPorosity[k][q];
-    }
-  } );
+void PorosityBase::initializeState() const
+{
+  m_porosity_n.setValues< parallelDevicePolicy<> >( m_newPorosity.toViewConst() );
+  m_initialPorosity.setValues< parallelDevicePolicy<> >( m_newPorosity.toViewConst() );
```

Comment:
Note you are launching two kernels here. Maybe we need a `setValues` that can perform a batch.

Reply:
Yes, that would be very useful to have a batch `setValues`. I can go back to the `forAll` loop in a future PR.

Reply:
Realistically though, what we need is an array copying interface that would dispatch to `memcpy`/`cudaMemcpy` when types/shapes match